LoongArch support patch V1

kilaterlee opened the pull request:
Make the EDK2 to support LoongArch architecture.

kilaterlee committed:
FatPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture for EDK2 CI testing.

Cc: Ray Ni <ray.ni@intel.com>

Signed-off-by: Chao Li  <lichao@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
FmpDevicePkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture for EDK2 CI testing.

Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Guomin Jiang <guomin.jiang@intel.com>
Cc: Wei6 Xu <wei6.xu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
NetworkPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture for EDK2 CI testing.

Cc: Maciej Rabeda <maciej.rabeda@linux.intel.com>
Cc: Jiaxin Wu <jiaxin.wu@intel.com>
Cc: Siyuan Fu <siyuan.fu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Jiaxin Wu <jiaxin.wu@intel.com>

kilaterlee committed:
SecurityPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture to SecurityPkg for EDK2 CI testing.

Cc: Jiewen Yao <jiewen.yao@intel.com>
Cc: Jian J Wang <jian.j.wang@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Jiewen Yao <Jiewen.yao@intel.com>

kilaterlee committed:
ShellPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture to ShellPkg for EDK2 CI testing.

Cc: Ray Ni <ray.ni@intel.com>
Cc: Zhichao Gao <zhichao.gao@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>
Reviewed-by: Zhichao Gao <zhichao.gao@intel.com>

kilaterlee committed:
UnitTestFrameworkPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture to UnitTestFramworkPkg for LOONGARCH64 EDK2
CI.

Cc: Michael D Kinney <michael.d.kinney@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/DxeServicesLib: Add LOONGARCH64 architecture

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture to MdePkg/DxeServiceLib.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdeModulePkg: Use LockBoxNullLib for LOONGARCH64

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

LoongArch doesn't have SMM by now.

Cc: Jian J Wang <jian.j.wang@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseIoLibIntrinsic: IoLibNoIo for LoongArch architecture.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

LoongArch MMIO library instance, use the IoLibNoIo.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseCpuLib: LoongArch Base CPU library implementation.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Implement LoongArch CPU related functions in BaseCpuLib.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseSafeIntLib: Add LoongArch64 architecture for BaseSafeIntLib.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 architecture for BaseSafeIntLib library.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdeModulePkg/Logo: Add LoongArch64 architecture.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 architecture to the Logo.

Cc: Zhichao Gao <zhichao.gao@intel.com>
Cc: Ray Ni <ray.ni@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>
Reviewed-by: Zhichao Gao <zhichao.gao@intel.com>

kilaterlee committed:
MdeModulePkg/CapsuleRuntimeDxe: Add LoongArch64 architecture.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch in INF for building CapsuleRuntimeDxe LoongArch64 image.

Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Guomin Jiang <guomin.jiang@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdeModulePkg/DxeIplPeim : LoongArch DxeIPL implementation.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Implement LoongArch DxeIPL instance.

Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Guomin Jiang <guomin.jiang@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
NetworkPkg: Add LoongArch64 architecture.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 architecture in to NetworkPkg.

Cc: Maciej Rabeda <maciej.rabeda@linux.intel.com>
Cc: Jiaxin Wu <jiaxin.wu@intel.com>
Cc: Siyuan Fu <siyuan.fu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
BaseTools: Add LoongArch64 binding.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 ProcessorBin.h and add LoongArch to Makefiles.

Cc: Bob Feng <bob.c.feng@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Yuwei Chen <yuwei.chen@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
BaseTools: Enable LoongArch64 architecture for LoongArch64 EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

EDK CI for LoongArch64 architecture

Enable LoongArch64 architecture for LoongArch64 EDK2 CI testing.

Cc: Bob Feng <bob.c.feng@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Yuwei Chen <yuwei.chen@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
.azurepipelines: Add LoongArch64 architecture on LoongArch64 EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 architecture on LoongArch64 EDK2 CI.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
.pytool: Add LoongArch64 architecture on LoongArch64 EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch64 architecture on LoongArch64 EDK2 CI testing.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg: Add LoongArch LOONGARCH64 binding

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 sections in MdePkg.dec and LOONGARCH64 ProcessorBind.h

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Co-authored-by: Dongyan Qian <qiandongyan@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BasePeCoff: Add LoongArch PE/Coff related code.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch image relocation.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
NetworkPkg/HttpBootDxe: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH architecture for EDK2 CI testing.

Cc: Maciej Rabeda <maciej.rabeda@linux.intel.com>
Cc: Jiaxin Wu <jiaxin.wu@intel.com>
Cc: Siyuan Fu <siyuan.fu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Jiaxin Wu <jiaxin.wu@intel.com>

kilaterlee committed:
CryptoPkg: Add LOONGARCH64 architecture for EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LOONGARCH64 architecture for EDK2 CI testing.

Cc: Jiewen Yao <jiewen.yao@intel.com>
Cc: Jian J Wang <jian.j.wang@intel.com>
Cc: Xiaoyu Lu <xiaoyu1.lu@intel.com>
Cc: Guomin Jiang <guomin.jiang@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Jiewen Yao <Jiewen.yao@intel.com>

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -267,7 +268,7 @@ DEFINE DTC_BIN                 = ENV(DTC_PREFIX)dtc
 #                               https://acpica.org/downloads
 #   GCC5        -Linux,Windows-  Requires:
-#                             GCC 5 with LTO support, targeting x86_64-linux-gnu, aarch64-linux-gnu, arm-linux-gnueabi or riscv64-linux-gnu
+#                             GCC 5 with LTO support, targeting x86_64-linux-gnu, aarch64-linux-gnu, arm-linux-gnueabi, riscv64-linux-gnu ro loongarch64-linux-gnu
```
```suggestion
#                             GCC 5 with LTO support, targeting x86_64-linux-gnu, aarch64-linux-gnu, arm-linux-gnueabi, riscv64-linux-gnu or loongarch64-linux-gnu

```

xen0n replied:
typo 还是没改

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -1870,19 +1871,23 @@ DEFINE GCC_ALL_CC_FLAGS            = -g -Os -fshort-wchar -fno-builtin -fno-stri
 DEFINE GCC_IA32_CC_FLAGS           = DEF(GCC_ALL_CC_FLAGS) -m32 -malign-double -freorder-blocks -freorder-blocks-and-partition -O2 -mno-stack-arg-probe
 DEFINE GCC_X64_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -mno-red-zone -Wno-address -mno-stack-arg-probe
 DEFINE GCC_ARM_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -mlittle-endian -mabi=aapcs -fno-short-enums -funsigned-char -ffunction-sections -fdata-sections -fomit-frame-pointer -Wno-address -mthumb -mfloat-abi=soft -fno-pic -fno-pie
+DEFINE GCC_LOONGARCH64_CC_FLAGS    = DEF(GCC_ALL_CC_FLAGS) -march=loongarch64 -mabi=lp64 -Wa,-mla-global-with-abs -fno-plt -Wno-address -fno-short-enums -fsigned-char -ffunction-sections -fdata-sections
```
这是老工具链？`-mabi` 取值在上游提交的版本基本一定会变成 `lp64s`（如果 UEFI 同内核一样不应当使用 FPU）或者 `lp64d`（如果可以使用 FPU），`lp64` 是不支持的，到时候还要再改一次。

kilaterlee replied:
对的，这个是用老工具链做的patch，新工具链上传后验证后我会再改一次

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -1870,19 +1871,23 @@ DEFINE GCC_ALL_CC_FLAGS            = -g -Os -fshort-wchar -fno-builtin -fno-stri
 DEFINE GCC_DLINK2_FLAGS_COMMON     = -Wl,--script=$(EDK_TOOLS_PATH)/Scripts/GccBase.lds
 DEFINE GCC_IA32_X64_DLINK_COMMON   = DEF(GCC_DLINK_FLAGS_COMMON) --gc-sections
 DEFINE GCC_ARM_AARCH64_DLINK_COMMON= -Wl,--emit-relocs -nostdlib -Wl,--gc-sections -u $(IMAGE_ENTRY_POINT) -Wl,-e,$(IMAGE_ENTRY_POINT),-Map,$(DEST_DIR_DEBUG)/$(BASE_NAME).map
+DEFINE GCC_LOONGARCH64_DLINK_COMMON= -Wl,--emit-relocs -nostdlib -Wl,--gc-sections -u $(IMAGE_ENTRY_POINT) -Wl,-e,$(IMAGE_ENTRY_POINT),-Map,$(DEST_DIR_DEBUG)/$(BASE_NAME).map
 DEFINE GCC_ARM_DLINK_FLAGS         = DEF(GCC_ARM_AARCH64_DLINK_COMMON) -z common-page-size=0x20 -Wl,--pic-veneer
 DEFINE GCC_AARCH64_DLINK_FLAGS     = DEF(GCC_ARM_AARCH64_DLINK_COMMON) -z common-page-size=0x20
+DEFINE GCC_LOONGARCH64_DLINK_FLAGS = DEF(GCC_LOONGARCH64_DLINK_COMMON) -z common-page-size=0x20
```
ARM/AArch64 的链接参数区分 common 和各自两个部分是因为两个架构需要的参数不完全相同，现在 LA 没有第二个架构，应该不需要区分

tivycc replied:
> ARM/AArch64 的链接参数区分 common 和各自两个部分是因为两个架构需要的参数不完全相同，现在 LA 没有第二个架构，应该不需要区分

common部分保留为好，这个架构是指指令集还是具体实现方式？

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -2463,6 +2475,35 @@ RELEASE_GCC5_AARCH64_DLINK_XIPFLAGS = -z common-page-size=0x20
+*_GCC5_LOONGARCH64_DLINK2_FLAGS         = DEF(GCC5_LOONGARCH64_DLINK2_FLAGS)
+*_GCC5_LOONGARCH64_RC_FLAGS             = DEF(GCC_LOONGARCH64_RC_FLAGS)
+*_GCC5_LOONGARCH64_OBJCOPY_FLAGS        =
+*_GCC5_LOONGARCH64_NASM_FLAGS           = -f elf32
```
这是啥，需要吗？

kilaterlee replied:
需要，因为EDK2现在最高支持到GCC5，虽然他不管你用什么版本的cross-compiler，但是这里必须写一个版本

xen0n replied:
啊，意思是这里 `NASM_FLAGS` 那个 `-f elf32` 感觉不到存在的意义。我不太熟悉固件的开发流程（只是看过代码），可以解释一下吗？

kilaterlee replied:
这个flags应该是延续之前的写法，我感觉去掉也没问题。

xen0n replied:
> 这个flags应该是延续之前的写法，我感觉去掉也没问题。

嗯，那看下别的新架构没加就去了吧，一方面 NASM 是 x86-only 的（当然我不知道他们有没有复用这个名字去调 `as`，正如 `GCC5` 不是真 gcc 5 一样），另一方面我们这次没做 32 位支持，放个 elf32 在这儿只会误导群众。

kilaterlee replied:
> > 这个flags应该是延续之前的写法，我感觉去掉也没问题。
> 
> 嗯，那看下别的新架构没加就去了吧，一方面 NASM 是 x86-only 的（当然我不知道他们有没有复用这个名字去调 `as`，正如 `GCC5` 不是真 gcc 5 一样），另一方面我们这次没做 32 位支持，放个 elf32 在这儿只会误导群众。

可以，我试试去掉有没有问题

xen0n commented on BaseTools/Source/C/GenFv/GenFvInternalLib.c:
```diff
@@ -2416,6 +2418,107 @@ Routine Description:
+
+  We support two schemes on LoongArch:
+  1) Beginning of the FV is the reset vector
+  2) Reset vector is data bytes FDF file and that code branches to reset vector
```
`Reset vector is data bytes FDF file` 英语读不通，啥叫“复位向量是数据字节 FDF 文件”……可以考虑改一下措辞。

另外上面 1 这一项的句尾没有标点，应该用分号或者逗号。

tivycc replied:
实际表达是”以字节序形式存在于FDF文件“，Exist with byteorder in the FDF file？

xen0n replied:
“字节序”也没看懂，感觉像是“字节流”。。这样一来整句话的汉语表述就是“复位向量以字节流形式存在于 FDF 文件，这些代码跳转到复位向量”，后半部分又读不通了。

如果前半句只是表达“存在于 FDF 文件”，那写 "stored in the FDF" 就行了。（我怀疑 FDF 的第二个 F 就是“文件”）

kilaterlee replied:
我修改一下

xen0n commented on BaseTools/Source/C/GenFv/GenFvInternalLib.c:
```diff
@@ -2416,6 +2418,107 @@ Routine Description:
+  2) Reset vector is data bytes FDF file and that code branches to reset vector
+    in the beginning of the FV (fixed size offset).
+
+  We relocate it to SecCoreEntry and copy the codes that in ResetVector to the beginning of the FV.
```
```suggestion
  We relocate it to SecCoreEntry and copy the ResetVector code to the beginning of the FV.

```

xen0n commented on BaseTools/Source/C/GenFv/GenFvInternalLib.c:
```diff
@@ -2416,6 +2418,107 @@ Routine Description:
+      InstrStack = (SecCoreEntryAddress & 0x0fff);
+      ResetVector[1] = 0x03800021 | (InstrStack << 10);   /* ori ra, ra, ui12 */
+      ResetVector[2] = 0x4c000021;                        /* jirl ra, ra, 0 */
+      ResetVector[3] = 0x00141000;                        /* and $r0, $r0, $r0, for nop */
```
LA 的跳转没有延迟槽，这个 nop 不会被执行到，这里加上是要满足什么对齐要求吗？如果没有必须加的原因的话建议去掉。

This aside，其他地方和手册描述不是都用 `0x03400000 = andi zero, zero, 0` 表示 `nop` 吗？建议所有涉及的地方都统一一下。

tivycc replied:
> LA 的跳转没有延迟槽，这个 nop 不会被执行到，这里加上是要满足什么对齐要求吗？如果没有必须加的原因的话建议去掉。
> 
> This aside，其他地方和手册描述不是都用 `0x03400000 = andi zero, zero, 0` 表示 `nop` 吗？建议所有涉及的地方都统一一下。

结构定义存在一个16字节的空域，最好还是填满

kilaterlee replied:
那四条指令跳走就行了，不用用满了吧

xen0n replied:
> 那四条指令跳走就行了，不用用满了吧

看起来跳走只需要 3 条指令（2 条制作 ra，1 条跳进 ra），@tivycc 意思估计是确保数组每个成员都有确定的值会降低心智负担？

kilaterlee replied:
> 

我觉得不用用满了，FV格式给了4个4字节就是自由发挥的，我们都已经跳走了，那么后面的指令就不会执行，除非考虑到攻击的情况，篡改FV header内容，那就是用满了也没辙

xen0n replied:
我意思反正就是如果决定不要填，那把这个 nop 去掉；如果决定要填，那把 nop 指令变成 `0x03400000` 保持跟惯例一致。别的可以不用管。。。

kilaterlee replied:
> 我意思反正就是如果决定不要填，那把这个 nop 去掉；如果决定要填，那把 nop 指令变成 `0x03400000` 保持跟惯例一致。别的可以不用管。。。

我是觉得可以不填，看看其他人的想法 @qiandongyan @bibo-mao

xen0n commented on BaseTools/Source/C/GenFw/Elf64Convert.c:
```diff
@@ -1417,6 +1423,10 @@ WriteSections64 (
+        } else if (mEhdr->e_machine == EM_LOONGARCH64) {
+            //
+            //do nothing
+            //
```
写清楚原因，另外 `//` 和注释内容之间需要有一个空格。

kilaterlee replied:
这里确实不需要做什么，如果这么写觉得不好，那就去掉

xen0n replied:
> 这里确实不需要做什么，如果这么写觉得不好，那就去掉

不是这意思，意思是如果你明确知晓在你的场景下，由于某些原因，“不做任何事情”是正确的、合适的，那就应当把这些原因写清楚，方便其他看代码的人知道为啥你场景特殊，为啥你啥都不做是合适的。

xen0n commented on BaseTools/Source/C/GenFw/Elf64Convert.c:
```diff
@@ -1647,6 +1657,71 @@ WriteRelocations64 (
+              case R_LARCH_SUB64:
+              case R_LARCH_GNU_VTINHERIT:
+              case R_LARCH_GNU_VTENTRY:
+                break;
```
写清楚不处理的原因。

kilaterlee replied:
Relocation type只用了一个MARK的type，见[PE规范](https://docs.microsoft.com/en-us/windows/win32/debug/pe-format) Base Relocation Types章节。

xen0n replied:
嗯，那跟我猜的一样，这里可以补一句 `// These relocation types are unused in PE format for LoongArch.` 解释一下（参考措辞，你们可以改掉）

kilaterlee replied:
> 嗯，那跟我猜的一样，这里可以补一句 `// These relocation types are unused in PE format for LoongArch.` 解释一下（参考措辞，你们可以改掉）

可以

xen0n commented on BaseTools/Source/C/Include/LoongArch64/ProcessorBind.h:
```diff
@@ -0,0 +1,80 @@
+/** @file
+  Processor or Compiler specific defines and types for LoongArch
```
`LoongArch64`?

kilaterlee replied:
我觉得很多场景可以不用强调64吧，因为这个文件夹就是64的==!

xen0n replied:
哦哦，这个叙述上保持跟 RV 之类新架构风格协调就行，只是问一下

xen0n commented on BaseTools/Source/C/Include/LoongArch64/ProcessorBind.h:
```diff
@@ -0,0 +1,80 @@
+//
+// The stack alignment required for LoongArch
+//
+#define CPU_STACK_ALIGNMENT   sizeof(UINT64)
```
确认下这边要的是硬件支持的最小对齐还是 ABI 要求的对齐？LA64 ABI 的栈应该是 16 字节对齐的，这里如果是要表达 ABI 对齐要求，那得改。

kilaterlee replied:
@qiandongyan 你确认一下这个

xen0n commented on BaseTools/Source/Python/UPT/Library/DataType.py:
```diff
@@ -463,6 +466,7 @@
 TAB_BINARIES_IPF = TAB_BINARIES + TAB_SPLIT + TAB_ARCH_IPF
 TAB_BINARIES_ARM = TAB_BINARIES + TAB_SPLIT + TAB_ARCH_ARM
 TAB_BINARIES_EBC = TAB_BINARIES + TAB_SPLIT + TAB_ARCH_EBC
+TAB_BINARIES_LOONGARCH64= TAB_BINARIES + TAB_SPLIT + TAB_ARCH_LOONGARCH64
```
```suggestion
TAB_BINARIES_LOONGARCH64 = TAB_BINARIES + TAB_SPLIT + TAB_ARCH_LOONGARCH64

```

xen0n commented on BaseTools/Source/Python/UPT/Library/DataType.py:
```diff
@@ -455,6 +457,7 @@
 TAB_SOURCES_IPF = TAB_SOURCES + TAB_SPLIT + TAB_ARCH_IPF
 TAB_SOURCES_ARM = TAB_SOURCES + TAB_SPLIT + TAB_ARCH_ARM
 TAB_SOURCES_EBC = TAB_SOURCES + TAB_SPLIT + TAB_ARCH_EBC
+TAB_SOURCES_LOONGARCH64 = TAB_SOURCES + TAB_SPLIT + TAB_ARCH_LOONGARCH64
```
前面都加在 EBC 前面，从这开始往后都是加在 EBC 后边（最后）了，统一一下

xen0n commented on MdePkg/Include/Library/BaseLib.h:
```diff
@@ -152,6 +153,29 @@ typedef struct {
+/// The LoongArch architecture context buffer used by SetJump() and LongJump()
+///
+typedef struct {
+  UINT64                            S0;
+  UINT64                            S1;
+  UINT64                            S2;
+  UINT64                            S3;
+  UINT64                            S4;
+  UINT64                            S5;
+  UINT64                            S6;
+  UINT64                            S7;
+  UINT64                            S8;
+  UINT64                            SP;
+  UINT64                            FP;
+  UINT64                            RA;
```
要这么多空格？

kilaterlee replied:
coding style

xen0n commented on MdePkg/Include/Protocol/DebugSupport.h:
```diff
@@ -654,17 +654,68 @@ typedef struct {
+  UINT64  PRMD;  // PRe-exception MoDe information
+  UINT64  EUEN;  // Extended component Unit ENable
+  UINT64  MISC;  // MISCellaneous controller
+  UINT64  ECFG;  // Excetption ConFiGuration
```
```suggestion
  UINT64  ECFG;  // Exception ConFiGuration

```

xen0n commented on MdePkg/Include/Protocol/DebugSupport.h:
```diff
@@ -654,17 +654,68 @@ typedef struct {
+  UINT64  ECFG;  // Excetption ConFiGuration
+  UINT64  ESTAT; // Exception STATus
+  UINT64  ERA;   // Exception Return Address
+  UINT64  BADV;  // Bad virtual Address
```
```suggestion
  UINT64  BADV;  // BAD Virtual address

```

xen0n commented on MdePkg/Include/Protocol/DebugSupport.h:
```diff
@@ -654,17 +654,68 @@ typedef struct {
+  UINT64  ESTAT; // Exception STATus
+  UINT64  ERA;   // Exception Return Address
+  UINT64  BADV;  // Bad virtual Address
+  UINT64  BADI;  // Bad Instruction
```
```suggestion
  UINT64  BADI;  // BAD Instruction

```

xen0n commented on MdePkg/Include/Uefi/UefiSpec.h:
```diff
@@ -2195,12 +2196,13 @@ typedef struct {
+#define EFI_REMOVABLE_MEDIA_FILE_NAME_ARM          L"\\EFI\\BOOT\\BOOTARM.EFI"
+#define EFI_REMOVABLE_MEDIA_FILE_NAME_AARCH64      L"\\EFI\\BOOT\\BOOTAA64.EFI"
+#define EFI_REMOVABLE_MEDIA_FILE_NAME_RISCV64      L"\\EFI\\BOOT\\BOOTRISCV64.EFI"
+#define EFI_REMOVABLE_MEDIA_FILE_NAME_LOONGARCH64  L"\\EFI\\BOOT\\BOOTLOONGARCH64.EFI"
```
以前出货的固件都检查的 `BOOTLOONGARCH.EFI`，通过 `strings` 检查固件二进制，没见有 `BOOTLOONGARCH64.EFI` 的支持。

现在这么写，从上游角度看倒是干净，不过到时候新固件配老系统、LiveUSB 的场景应该是启动不了的，那么打算怎么兼容之前的系统呢？

tivycc replied:
老系统是个问题，LiveUSB可以把BOOTLOONGARCH.EFI做个软连接到BOOTLOONGARCH.64EFI，安装完后系统会自己生成一个启动项。这里只是指定默认搜索项

xen0n replied:
澄清：我无意要求推上游的东西（规范文本也好，参考实现也好）一定要支持早期历史做法，这里只是提醒大家思考这个问题（如果有的人没考虑到这一块的话）。如果都想过了，那自然是最好的。

kilaterlee replied:
> 老系统是个问题，LiveUSB可以把BOOTLOONGARCH.EFI做个软连接到BOOTLOONGARCH.64EFI，安装完后系统会自己生成一个启动项。这里只是指定默认搜索项

只能推动现在开始做链接开始兼容，我已经发出去了，商业OS开始做了

xen0n replied:
行，上游不带历史包袱是最好的。

xen0n replied:
行，上游不带历史包袱是最好的。

xen0n commented on MdePkg/Library/BaseCacheMaintenanceLib/LoongArchCache.c:
```diff
@@ -0,0 +1,241 @@
+  }
+
+  ASSERT ((Length - 1) <= (MAX_ADDRESS - (UINTN)Address));
+  return Address;
```
这是空操作吗？写一下原因，明确下是目前没做还是 LA 这个操作不需要软件显式配合？

xen0n commented on MdePkg/Library/BaseCacheMaintenanceLib/LoongArchCache.c:
```diff
@@ -0,0 +1,241 @@
+  VOID
+  )
+{
+  DEBUG((DEBUG_ERROR, "%a:LoongArch unsupported function now.\n", __FUNCTION__));
```
建议修一下英语：

```suggestion
  DEBUG((DEBUG_ERROR, "%a: Not implemented on LoongArch for now.\n", __FUNCTION__));

```

下同。

tivycc replied:
这里应该既不是不支持也不是没实现，而是不需要的意思吧

xen0n replied:
> 这里应该既不是不支持也不是没实现，而是不需要的意思吧

如果是不需要的意思，那应该去掉 log message，用注释说明为啥 LA 上空操作是适合的。

kilaterlee replied:
> > 这里应该既不是不支持也不是没实现，而是不需要的意思吧
> 
> 如果是不需要的意思，那应该去掉 log message，用注释说明为啥 LA 上空操作是适合的。

这个可以实现，后面我加上去，用ibar实现

xen0n replied:
> > > 这里应该既不是不支持也不是没实现，而是不需要的意思吧
> > 
> > 
> > 如果是不需要的意思，那应该去掉 log message，用注释说明为啥 LA 上空操作是适合的。
> 
> 这个可以实现，后面我加上去，用ibar实现

🉑

xen0n commented on MdePkg/Library/BaseLib/LoongArch64/CpuPause.S:
```diff
@@ -0,0 +1,31 @@
+#**/
+
+ASM_PFX(CpuPause):
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
+    andi    $zero, $zero, 0x0  //nop
```
这边直接写 nop 有何不可？

xen0n replied:
这问题没回复来着？

kilaterlee replied:
我觉得都可以，nop在LoongArch里面是宏指令，也是这么完成的，所以就没改，我觉得也没必要改

xen0n replied:
不是，因为有更简明表意的写法，所以建议是用，跟编译结果无关。如果讲编译结果一样就不需要改的话，那我 move 都可以不用，反正编出来也是 `or rd, rj, zero`...

因为现在看这段汇编就像是来自没有伪指令支持的远古时期（我猜的，去年年初最早能看到的 LoongArch 汇编全是这样的，没有 nop 没有 jr，有的连 move 都没有，最多只有 li），跟近几个月以来常见的 LoongArch 汇编写作风格已经不搭了。

kilaterlee replied:
X86上这个函数的原意是pause指令，空闲40拍取指；我这里在考虑是不是用ibar 0来做，如果用ibar 0，我需要测试一下

kilaterlee replied:
已经改成nop了

xen0n commented on MdePkg/Library/BaseLib/LoongArch64/GetInterruptState.S:
```diff
@@ -0,0 +1,35 @@
+#  @retval TRUE  CPU interrupts are enabled.
+#  @retval FALSE CPU interrupts are disabled.
+#
+#**/
+
+ASM_PFX(GetInterruptState):
+  li.w  $t1, 0x4
+  csrrd $t0, 0x0
+  and   $t0, $t0, $t1
+  beqz  $t0, 1f
+  li.w  $v0, 0x1
```
少用 `v0/v1` 写法吧，工具链那边已经[不建议使用](https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-CN.html)了。都改 `a0/a1` 就行。

kilaterlee replied:
好的，这部分改掉

xen0n commented on MdePkg/Include/Library/BaseLib.h:
```diff
@@ -152,6 +153,29 @@ typedef struct {
+  UINT64                            S7;
+  UINT64                            S8;
+  UINT64                            SP;
+  UINT64                            FP;
```
如果 UEFI 代码不强制要求在 `fp` 维护一个正确的栈帧指针，那最好把 `fp` 当作 `s9`，这样看着表意清晰一些（本质上不作帧指针的 `fp` 就是 `s9` 而已，又或者说，与其说 `fp` 偶尔可以当 `s9` 不如说 `s9` 偶尔可以作 `fp`）。这也是新的 [ABI 规范](https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-CN.html) 这么写的。

kilaterlee replied:
在platform代码里面FP确实用了，这个用于完成跳转的方案，具体我再看看代码

xen0n commented on MdePkg/Library/BaseLib/LoongArch64/SwitchStack.S:
```diff
@@ -0,0 +1,39 @@
+**/
+
+ASM_PFX(InternalSwitchStackAsm):
+    LOAD    $ra, $a0, RSIZE *11
```
乘号之后没加空格，前面一个文件倒加了，建议这边也加下

xen0n commented on MdePkg/Library/BaseSynchronizationLib/LoongArch64/Synchronization.c:
```diff
@@ -0,0 +1,161 @@
+  )
+{
+  return *Value != CompareValue ? *Value :
+           ((*Value = ExchangeValue), CompareValue);
```
这个写法不对吧，这能有原子性？看注释这里应该必须有原子 CAS 语义才对。

xen0n commented on MdePkg/Library/BaseSynchronizationLib/LoongArch64/Synchronization.c:
```diff
@@ -0,0 +1,161 @@
+    __asm__ __volatile__(
+      "ll.d  %0, %1    \n"
+      "addi.d  %0, %0,  1 \n"
+      "sc.d  %0, %1    \n"
```
多余的空格不去掉？现在完全看不出这些空格是出于何种对齐的考虑。

tivycc replied:
> 多余的空格不去掉？现在完全看不出这些空格是出于何种对齐的考虑。

函数参数之间的空格倒是UEFI里面的代码风格

xen0n reviewed (changes requested):
整体上：

1. EDK2 上游对于 commit messages 标题行不使用句尾标点，请把 commit messages 标题行最后的句号全部去掉。
2. 自然语言中不属于标识符写法的 `LOONGARCH` 应该都变成 `LoongArch`，请自行排查并修改掉。

tivycc commented on .pytool/Plugin/SpellCheck/cspell.base.yaml:
```diff
@@ -178,6 +178,7 @@
         "msvcrtd",
-        "XIPFLAGS"
+        "XIPFLAGS",
+        "LOONGARCH"
```
这里不是LOONGARCH64？整篇应该区分LOONGARCH、LOONGARCH64、LOONGARCH32

xen0n replied:
RV 没区分，看了下代码，他们也有 RISCV64 这种写法，不知道这个拼写检查插件的行为如何。。。不过再怎么说这里也有写 `LOONGARCH` 还是 `LoongArch` 还是 `loongarch` 有没有区别的另一个问题。

kilaterlee replied:
这个按照RV的风格写的，这个拼写插件还得仔细看看代码如何工作的

kilaterlee replied:
这个确认应该是CI的时候识别，编译架构的时候需要识别

tivycc commented on MdePkg/Library/BasePeCoffLib/BasePeCoff.c:
```diff
@@ -1,6 +1,6 @@
 /** @file
   Base PE/COFF loader supports loading any PE32/PE32+ or TE image, but
-  only supports relocating IA32, x64, IPF, ARM, RISC-V and EBC images.
+  only supports relocating IA32, x64, IPF, ARM, RISC-V, LoongArch and EBC images.
```
这里最好改成LOONGARCH

bibo-mao commented on MdePkg/Library/BaseSynchronizationLib/LoongArch64/Synchronization.c:
```diff
@@ -0,0 +1,161 @@
+      :"=&r" (temp), "+m" (*Value)
+    );
+  }while (!temp);
+  return *Value;
```
为什么不用atomic 指令使用 ll/sc 方式？

xen0n replied:
我猜这代码可能是祖传的 MIPS UEFI port 改的，最早写出来的时候连 MIPS 龙芯都没 AMO，现在可能要重新评估一下

kilaterlee replied:
说实话，我觉得自增自减在单核系统里面，是不是没必要做成atom的？

kilaterlee replied:
但是虚拟机里面牵扯到OS里面，是不是需要得调研，物理机我认为可能是没必要做成atom的

xen0n replied:
> 说实话，我觉得自增自减在单核系统里面，是不是没必要做成atom的？

单核系统？为啥会假定这份代码只跑在单核上，因为 UEFI 不初始化 SMP 吗？比较好奇

kilaterlee replied:
> > 说实话，我觉得自增自减在单核系统里面，是不是没必要做成atom的？
> 
> 单核系统？为啥会假定这份代码只跑在单核上，因为 UEFI 不初始化 SMP 吗？比较好奇

因为大部分时间都是在单核跑代码，MP用的很少；据我了解Intel在服务器MC初始化的时候用了MP，其他时间都是单核跑；我们在platform里面用了MP，但是实现都是代码隔离，不会出现临界区这种东西

xen0n replied:
我觉得移植时候对于这种可能出现的下游使用姿势，还是尽量按照接口约定去做；不能说在我们的场景下那部分代码没走到，所以有的接口约定就可以不去管了。会坑到自己不说，其他拿这份代码二次开发的人不知道具体实现和接口文档不一致，也会受影响。

bibo-mao commented on MdePkg/Library/BaseSynchronizationLib/LoongArch64/Synchronization.c:
```diff
@@ -0,0 +1,161 @@
+    __asm__ __volatile__(
+      "ll.d  %0, %1    \n"
+      "addi.d  %0, %0,  1 \n"
+      "sc.d  %0, %1    \n"
```
为什么不用atomic 指令使用 ll/sc 方式？ 有loongarch架构需要uefi支持不支持原子指令吗

kilaterlee committed:
BaseTools: Updated build tools to support new LoongArch.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Python code changes for building EDK2 LoongArch platform.

Cc: Bob Feng <bob.c.feng@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Yuwei Chen <yuwei.chen@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

xen0n commented on BaseTools/Bin/gcc_loongarch64_unknown_linux_ext_dep.yaml:
```diff
@@ -0,0 +1,22 @@
+  "type": "web",
+  "name": "gcc_loongarch64_unknown_linux",
+  "source": "https://github.com/loongson/build-tools/releases/download/2021.12.21/loongarch64-clfs-2021-12-18-cross-tools-gcc-full.tar.xz",
+  "version": "12.0.0 20210810",
```
这里可以更新下，另外 version 这个字符串也同步下

kilaterlee replied:
这个版本是新世界最新的了，version我会在下一次push的时候修掉

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -1897,6 +1902,7 @@ DEFINE GCC_X64_RC_FLAGS            = -I binary -O elf64-x86-64        -B i386
 DEFINE GCC_ARM_RC_FLAGS            = -I binary -O elf32-littlearm     -B arm     --rename-section .data=.hii
 DEFINE GCC_AARCH64_RC_FLAGS        = -I binary -O elf64-littleaarch64 -B aarch64 --rename-section .data=.hii
 DEFINE GCC_RISCV64_RC_FLAGS        = -I binary -O elf64-littleriscv   -B riscv   --rename-section .data=.hii
+DEFINE GCC_LOONGARCH64_RC_FLAGS    = -I binary -O elf64-loongarch -B loongarch64 --alt-elf-eflags=0x3 --rename-section .data=.hii
```
`-B` 跟上面对齐吧，别的不强求

xen0n commented on BaseTools/Conf/tools_def.template:
```diff
@@ -1871,19 +1872,23 @@ DEFINE GCC_ALL_CC_FLAGS            = -g -Os -fshort-wchar -fno-builtin -fno-stri
 DEFINE GCC_IA32_CC_FLAGS           = DEF(GCC_ALL_CC_FLAGS) -m32 -malign-double -freorder-blocks -freorder-blocks-and-partition -O2 -mno-stack-arg-probe
 DEFINE GCC_X64_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -mno-red-zone -Wno-address -mno-stack-arg-probe
 DEFINE GCC_ARM_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -mlittle-endian -mabi=aapcs -fno-short-enums -funsigned-char -ffunction-sections -fdata-sections -fomit-frame-pointer -Wno-address -mthumb -mfloat-abi=soft -fno-pic -fno-pie
+DEFINE GCC_LOONGARCH64_CC_FLAGS    = DEF(GCC_ALL_CC_FLAGS) -march=loongarch64 -mabi=lp64d -Wa,-mla-global-with-abs -fno-plt -Wno-address -fno-short-enums -fsigned-char -ffunction-sections -fdata-sections
```
确认下 UEFI 真会用到浮点操作？如果不会的话，建议还是走 LP64S ABI，下面有个 `--alt-elf-eflags` 也同步改掉

kilaterlee replied:
会用到，有openssl的需求

xen0n replied:
> 会用到，有openssl的需求

Ack

xen0n commented on BaseTools/Source/C/Common/PeCoffLoaderEx.c:
```diff
@@ -332,3 +333,78 @@ PeCoffLoaderRelocateArmImage (
+
+  switch (RelocType) {
+    case EFI_IMAGE_REL_BASED_LOONGARCH64_MARK_LA:
+      /* The next four instructions are used to load a 64 bit address, we change it together*/
```
```suggestion
      /* The next four instructions are used to load a 64 bit address, relocate all of them */

```

xen0n commented on BaseTools/Source/C/Common/PeCoffLoaderEx.c:
```diff
@@ -332,3 +333,78 @@ PeCoffLoaderRelocateArmImage (
+    case EFI_IMAGE_REL_BASED_LOONGARCH64_MARK_LA:
+      /* The next four instructions are used to load a 64 bit address, we change it together*/
+      Value = (*(UINT32*)Fixup & 0x1ffffe0) << 7 |       /* lu12i.w 20bits from bit5 */
+              (*((UINT32*)Fixup + 1) & 0x3ffc00) >> 10;  /* ori  12bits from bit10 */
```
```suggestion
              (*((UINT32*)Fixup + 1) & 0x3ffc00) >> 10;  /* ori     12bits from bit10 */

```

xen0n commented on BaseTools/Plugin/LinuxGcc5ToolChain/LinuxGcc5ToolChain.py:
```diff
@@ -5,6 +5,7 @@
 #
 # Copyright (c) Microsoft Corporation
 # Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on BaseTools/Bin/gcc_loongarch64_unknown_linux_ext_dep.yaml:
```diff
@@ -0,0 +1,22 @@
+# This is only downloaded when a build activates scope gcc_loongarch64_unknown_linux
+#
+# Copyright (c) Microsoft Corporation.
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>
```

xen0n commented on .azurepipelines/Ubuntu-GCC5.yml:
```diff
@@ -3,6 +3,7 @@
 #
 # Copyright (c) Microsoft Corporation.
 # Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on BaseTools/Source/C/GenFv/GenFvInternalLib.c:
```diff
@@ -3450,6 +3561,11 @@ Routine Description:
         VerboseMsg("Located ARM/AArch64 SEC/PEI core in child FV");
         mArm = TRUE;
       }
+      // machine type is LOONGARCH64, set a flag so LOONGARCH64 reset vector procesing occurs
```
```suggestion
      // machine type is LOONGARCH64, set a flag so LoongArch64 reset vector is processed

```

xen0n commented on BaseTools/Source/C/GenFv/GenFvInternalLib.c:
```diff
@@ -3450,6 +3561,11 @@ Routine Description:
       }
+      // machine type is LOONGARCH64, set a flag so LOONGARCH64 reset vector procesing occurs
+      if ((MachineType == EFI_IMAGE_MACHINE_LOONGARCH64)) {
+        VerboseMsg("Located LOONGARCH64 SEC core in child FV");
```
```suggestion
        VerboseMsg("Located LoongArch64 SEC core in child FV");

```

xen0n commented on BaseTools/Source/C/GenFw/Elf64Convert.c:
```diff
@@ -1149,10 +1155,10 @@ WriteSections64 (
           //
-          // Skip error on EM_RISCV64 becasue no symble name is built
-          // from RISC-V toolchain.
+          // Skip error on EM_RISCV64 and EM_LOONGARCH64 becasue no symble name is built
```
顺便改下 typo

```suggestion
          // Skip error on EM_RISCV64 and EM_LOONGARCH64 because no symbol name is built

```

xen0n commented on BaseTools/Source/C/GenFw/Elf64Convert.c:
```diff
@@ -1417,6 +1423,74 @@ WriteSections64 (
+          case R_LARCH_GNU_VTINHERIT:
+          case R_LARCH_GNU_VTENTRY:
+            //
+            // These types are not used or do not need to fix the offsets.
```
```suggestion
            // These types are not used or do not require fixup.

```

xen0n commented on BaseTools/Source/C/GenFw/elf_common.h:
```diff
@@ -181,6 +182,7 @@ typedef struct {
 #define EM_AARCH64  183  /* ARM 64bit Architecture */
 #define EM_RISCV64  243 /* 64bit RISC-V Architecture */
 #define EM_RISCV    244 /* 32bit RISC-V Architecture */
+#define EM_LOONGARCH64 258 /* LoongArch 64-bit Architecture */
```
32-bit LoongArch is also 258, so this should be just EM_LOONGARCH? The `EM_RISCV = 244` is not present in upstream binutils either (RISCV is unconditionally 243 there).

kilaterlee replied:
Okay, it is really not appropriate to defined with EM_LOONGARCH64, I will refer spelling in binutils and update this and other related files. Thanks.

kilaterlee replied:
> Okay, it is really not appropriate to defined with EM_LOONGARCH64, I will refer spelling in binutils and update this and other related files. Thanks.

I checkout the code, EDK II BastTools distinguishes 32bit and 64bit, is it better to define two? Such as:
EM_LOONGARCH32 258
EM_LOONGARCH64 258
@ChenghuaXu Can you give some recommendation?
Refer EDKII code:https://github.com/tianocore/edk2/blob/master/BaseTools/Source/C/GenFw/Elf32Convert.c and https://github.com/tianocore/edk2/blob/master/BaseTools/Source/C/GenFw/Elf64Convert.c

kilaterlee replied:
The newest commit solves this issue.

xen0n commented on MdeModulePkg/Core/DxeIplPeim/LoongArch64/DxeLoadFunc.c:
```diff
@@ -0,0 +1,61 @@
+  TopOfStack = (VOID *) ((UINTN) BaseOfStack + EFI_SIZE_TO_PAGES (STACK_SIZE) * EFI_PAGE_SIZE - CPU_STACK_ALIGNMENT);
+  TopOfStack = ALIGN_POINTER (TopOfStack, CPU_STACK_ALIGNMENT);
+  //
+  // End of PEI phase singal
```
```suggestion
  // End of PEI phase signal

```

xen0n commented on MdePkg/Include/IndustryStandard/Dhcp.h:
```diff
@@ -259,24 +260,28 @@ typedef enum {
+#define PXE_CLIENT_ARCH_RISCV64       0x001B          /// RISC-V uefi 64 for PXE
+#define PXE_CLIENT_ARCH_RISCV128      0x001D          /// RISC-V uefi 128 for PXE
+#define PXE_CLIENT_ARCH_LOONGARCH32   0x0025          /// LOONGARCH uefi 32 for PXE
+#define PXE_CLIENT_ARCH_LOONGARCH64   0x0027          /// LOONGARCH uefi 64 for PXE
```
```suggestion
#define PXE_CLIENT_ARCH_LOONGARCH64   0x0027          /// LoongArch uefi 64 for PXE
```

xen0n commented on MdePkg/Include/IndustryStandard/Dhcp.h:
```diff
@@ -259,24 +260,28 @@ typedef enum {
+#define HTTP_CLIENT_ARCH_RISCV64      0x001C          /// RISC-V uefi 64 boot from http
+#define HTTP_CLIENT_ARCH_RISCV128     0x001E          /// RISC-V uefi 128 boot from http
+#define HTTP_CLIENT_ARCH_LOONGARCH32  0x0026          /// LOONGARCH uefi 32 boot from http
+#define HTTP_CLIENT_ARCH_LOONGARCH64  0x0028          /// LOONGARCH uefi 64 boot from http
```
```suggestion
#define HTTP_CLIENT_ARCH_LOONGARCH64  0x0028          /// LoongArch uefi 64 boot from http
```

xen0n commented on MdePkg/Library/BasePeCoffLib/LoongArch/PeCoffLoaderEx.c:
```diff
@@ -0,0 +1,132 @@
+  switch (RelocType) {
+    case EFI_IMAGE_REL_BASED_LOONGARCH64_MARK_LA:
+      Value = (*(UINT32*)Fixup & 0x1ffffe0) << 7 |       /* lu12i.w 20bits from bit5 */
+              (*((UINT32*)Fixup + 1) & 0x3ffc00) >> 10;  /* ori  12bits from bit10 */
```
```suggestion
              (*((UINT32*)Fixup + 1) & 0x3ffc00) >> 10;  /* ori     12bits from bit10 */

```

xen0n commented on MdePkg/Library/BaseSafeIntLib/BaseSafeIntLib.inf:
```diff
@@ -4,9 +4,10 @@
 # Copyright (c) 2017, Microsoft Corporation
-#  Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+# Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on MdePkg/Library/BaseSynchronizationLib/BaseSynchronizationLib.inf:
```diff
@@ -4,6 +4,7 @@
 #  Copyright (c) 2007 - 2018, Intel Corporation. All rights reserved.<BR>
 #  Portions copyright (c) 2008 - 2009, Apple Inc. All rights reserved.<BR>
 #  Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+#  Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
#  Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on MdePkg/Library/BaseSynchronizationLib/LoongArch64/Synchronization.c:
```diff
@@ -0,0 +1,239 @@
+/** @file
+  LoongArch synchronization functions.
+
+  Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
  Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on MdePkg/MdePkg.dec:
```diff
@@ -7,6 +7,7 @@
 # Copyright (c) 2007 - 2021, Intel Corporation. All rights reserved.<BR>
 # Portions copyright (c) 2008 - 2009, Apple Inc. All rights reserved.<BR>
 # (C) Copyright 2016 - 2021 Hewlett Packard Enterprise Development LP<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on MdePkg/MdePkg.dsc:
```diff
@@ -4,6 +4,7 @@
 # Copyright (c) 2007 - 2021, Intel Corporation. All rights reserved.<BR>
 # Portions copyright (c) 2008 - 2009, Apple Inc. All rights reserved.<BR>
 # (C) Copyright 2020 Hewlett Packard Enterprise Development LP<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

xen0n commented on NetworkPkg/Network.dsc.inc:
```diff
@@ -9,6 +9,7 @@
 #
 # Copyright (c) 2019 - 2020, Intel Corporation. All rights reserved.<BR>
 # Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
+# Copyright (c) 2021 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
```suggestion
# Copyright (c) 2021, Loongson Technology Corporation Limited. All rights reserved.<BR>

```

kilaterlee committed:
.python/SpellCheck: Add "Loongson" and "LOONGARCH" to "words" section

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add Loongson and LOONGARCH to "words" section in cspell.base.yaml file
to avoid spelling check error.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/Include: LoongArch definitions.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch processor related definitions.

For the Http boot and PXE boot types seeing this URL section "Processor
Architecture Type" for the LOONGARCH values:
https://www.iana.org/assignments/dhcpv6-parameters/dhcpv6-parameters.xhtml

For definitions of PE/COFF and LOONGARCH relocation types, see the
"Machine Types" and "Basic Relocation Types" sections of this URL for
LOONGARCH values:
https://docs.microsoft.com/en-us/windows/win32/debug/pe-format

For the register definitions of exceptions context, see the UEFI V2.10
18.2.2, 18.2.4 and 18.2.5 sections of this URL for LOONGARCH
definitions:
https://uefi.org/specs/UEFI/2.10/18_Protocols_Debugger_Support.html

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
BaseTools: Updated for GCC5 tool chain for LoongArch platfrom.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

BaseTools define template files changes for building EDK2 LoongArch
platform.

Cc: Bob Feng <bob.c.feng@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Yuwei Chen <yuwei.chen@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Dongyan Qian <qiandongyan@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
BaseTools: Update GenFw/GenFv to support LoongArch platform.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

C code changes for building EDK2 LoongArch platform.

For definitions of PE/COFF and LOONGARCH relocation types, see the
"Machine Types" and "Basic Relocation Types" sections of this URL for
LOONGARCH values:
https://docs.microsoft.com/en-us/windows/win32/debug/pe-format

Cc: Bob Feng <bob.c.feng@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Yuwei Chen <yuwei.chen@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Dongyan Qian <qiandongyan@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Co-authored-by: Yang Zhou <zhouyang@loongson.cn>
Co-authored-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Liming Gao <gaoliming@byosoft.com.cn>

kilaterlee committed:
MdePkg: Added file of DebugSupport.h to MdePkg.ci.yaml

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

DebugSupport.h is all defined by UEFI Spec, most of the code
doesn't fit EDKII coding style, add it to IgnoreFiles field to
make CI ECC check pass.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg: Added LoongArch jump buffer register definition to MdePkg.ci.yaml

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

If the new Arch register is defined in BaseLib.h when running
the CI tests, it will give an ECC check error. Add the
LoongArch register defined in the ExceptionList field to make
the CI ECC check pass.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/Include: Add LOONGARCH related definitions EDK2 CI.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

HTTP/PXE boot LOONGARCH64 related definitions for EDK2 CI.

For the LOONGARCH values, please seeing following URL section
"Processor Architecture Types":
https://www.iana.org/assignments/dhcpv6-parameters/dhcpv6-parameters.xhtml

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseLib: BaseLib for LOONGARCH64 architecture.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Add LoongArch LOONGARCH64 BaseLib functions.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseCacheMaintenanceLib: LoongArch cache maintenance implementation.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Implement LoongArch cache maintenance functions in
BaseCacheMaintenanceLib.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee committed:
MdePkg/BaseSynchronizationLib: LoongArch cache related code.

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4053

Support LoongArch cache related functions.

Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Liming Gao <gaoliming@byosoft.com.cn>
Cc: Zhiguang Liu <zhiguang.liu@intel.com>

Signed-off-by: Chao Li <lichao@loongson.cn>
Co-authored-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Reviewed-by: Michael D Kinney <michael.d.kinney@intel.com>

kilaterlee commented:
EDK2 has been merged, closed.
